pkg/executor/adapter.go

```diff
@@ -549,9 +549,13 @@ func (a *ExecStmt) Exec(ctx context.Context) (_ sqlexec.RecordSet, err error) {
 		_, planDigest := GetPlanDigest(stmtCtx)
 		_, sqlDigest := stmtCtx.SQLDigest()
 		stmtCtx.RunawayChecker = rm.DeriveChecker(stmtCtx.ResourceGroupName, stmtCtx.OriginalSQL, sqlDigest.String(), planDigest.String(), sessionVars.StartTime)
-		if err := stmtCtx.RunawayChecker.BeforeExecutor(); err != nil {
+		switchGroupName, err := stmtCtx.RunawayChecker.BeforeExecutor()
+		if err != nil {
 			return nil, err
 		}
+		if len(switchGroupName) > 0 {
+			stmtCtx.ResourceGroupName = switchGroupName
```

Comment:
Should we be concerned when a user inputs the wrong group name?

Reply:
We should, but this PR is already quite large. I've added a TODO here and will implement it in the next PR.

https://github.com/pingcap/tidb/pull/54804/files#diff-188b7b4333b4d28ccce2a7a470048f68c490ae669fc61ed87baf764069f96a14R57